packages/jellyfish-buffer/__tests__/Composer.test.ts

```diff
@@ -1182,6 +1182,82 @@ describe('ComposableBuffer.varUInt', () => {
   })
 })
 
+describe('ComposableBuffer.varInt', () => {
+  let value: number = 0
+  const composer = ComposableBuffer.varInt(() => value, (v: number) => value = v)
+
+  describe('1 byte = 0', () => {
+    it('should fromBuffer', () => {
+      shouldFromBuffer(composer, '00', 0x00, () => value)
+    })
+
+    it('should toBuffer', () => {
+      shouldToBuffer(composer, '00', 0x00, v => value = v)
+    })
+  })
+
+  describe('1 byte = 127', () => {
+    it('should fromBuffer', () => {
+      shouldFromBuffer(composer, '7f', 127, () => value)
+    })
+
+    it('should toBuffer', () => {
+      shouldToBuffer(composer, '7f', 127, v => value = v)
+    })
+  })
+
+  describe('2 byte = 128', () => {
+    it('should fromBuffer', () => {
+      shouldFromBuffer(composer, '8000', 128, () => value)
+    })
+
+    it('should toBuffer', () => {
+      shouldToBuffer(composer, '8000', 128, v => value = v)
+    })
+  })
+
+  describe('2 bytes = 256', () => {
+    it('should fromBuffer', () => {
+      const buffer = new SmartBuffer()
+      buffer.writeUInt8(0x81)
+      buffer.writeUInt8(0x00)
+
+      composer.fromBuffer(buffer)
+      expect(value).toStrictEqual(256)
+    })
+
+    it('should toBuffer', () => {
+      value = 256
+
+      const buffer = new SmartBuffer()
+      composer.toBuffer(buffer)
+
+      const expected = Buffer.from('8100', 'hex')
+      expect(buffer.toString()).toStrictEqual(expected.toString())
+    })
+  })
+
+  describe('2 byte = 16511', () => {
+    it('should fromBuffer', () => {
+      shouldFromBuffer(composer, 'ff7f', 16511, () => value)
```

Comment:
@fuxingloh  sorry i don't know how it works
eg:
```sh
echo $((0x7f)) # output: 127
echo $((0xff7f) # output: 65407
```
how does `ff7f` == 16511??

Reply:
So VarInt is a 128-bit operation, think of it as there is a bit operation every 128 bits, 128 in unsigned decimal or up to 0x7f in hex.

> Variable-length integers: bytes are a MSB base-128 encoding of the number. **The high bit in each byte signifies whether another digit follows.** To make sure the encoding is one-to-one, one is subtracted from all but the last digit. Thus, the byte sequence a[] with length len, where all but the last byte has bit 128 set, encodes the number:

So for simplification think:

1. `0` is `0x00`
2. `1` is `0x01`
3. ... and `127` is `0x7f`
4. and now `128` there are 2 "128 bits", hence presented in HEX it's `0x8000`, 
5. `129` is `0x8001`

When reading the VarInt buffer array [`0x01`], `0x01` is a low "byte"; <128. Hence there isn't another digit that follows. Thus just `1`.

When reading the VarInt buffer array [`0x80`, `0x01`] , `0x80` is a high "byte"; >128. Hence there is another digit that follows, read `0x01`, okay that's a low byte. `0x80 %  0x80 * 0x80 + 0x01` = 129...

Or in bitwise operations:

```ts
export function readVarInt (buffer: SmartBuffer): number {
  let n = 0

  while (true) {
    const buff: number = buffer.readUInt8()
    n = n << 7
    n += buff & 0x7f
    if ((buff & 0x80) !== 0) {
      n = n += 1
    } else {
      return n
    }
  }
}
```

// Breaking it down: 🥲
```ts
// init
let n = 0
const buffer = [0x8001]

// buff=0x80
let buff: number = buffer.readUInt8()

// shift left n by 7 bits
// n = 0 // still 0
n = n << 7

// n = 0 + (0x80 & 0x7f)
// n = 0 // still 0
n = n + (buff & 0x7f)

// (0x80 & 0x80) !== 0, true
// n = 0 + 1, n = 1
if ((buff & 0x80) !== 0) {
  n = n + 1
} else {
  return n
}

// buff=0x01
let buff: number = buffer.readUInt8()

// shift left n by 7 bits,  1 << 7 = 128
// n = 128
n = n << 7

// n = 128 + (0x01 & 0x7f)
// n = 129
n = n + (buff & 0x7f)

// (0x01 & 0x80) !== 0, false
// return 129
if ((buff & 0x80) !== 0) {
  n = n + 1
} else {
  return n
}
```

Reply:
i won't resolve it.. this thread is useful.. just let it open